tests/unit/test_checksum.cpp

```diff
@@ -0,0 +1,280 @@
+
+        // Verify that the calculated UDP checksum is valid.
+
+        VerifyOrQuit(
+            CalculateChecksum(messageInfo.GetSockAddr(), messageInfo.GetPeerAddr(), Ip6::kProtoUdp, *message) == 0xffff,
+            "UDP checksum does not match expected value");
+
+        // Verify that `Checksum::VerifyMessageChecksum()` accepts the
+        // message and its calculated checksum.
+
+        SuccessOrQuit(Checksum::VerifyMessageChecksum(*message, messageInfo, Ip6::kProtoUdp),
+                      "Checksum::VerifyMessageChecksum() failed");
+
+        message->Free();
+    }
+}
+
+void TestIcmp6MessageChecksum(void)
+{
+    enum : uint16_t
+    {
+        kMinSize = sizeof(Ip6::Icmp::Header),
+        kMaxSize = kBufferSize * 3 + 24,
+    };
+
+    const char *kSourceAddress = "fd00:feef:dccd:baab:9889:7667:5444:3223";
+    const char *kDestAddress   = "fd01:abab:beef:cafe:1234:5678:9abc:0";
+
+    Instance *instance = static_cast<Instance *>(testInitInstance());
+
+    VerifyOrQuit(instance != nullptr, "Null OpenThread instance\n");
+
+    for (uint16_t size = kMinSize; size <= kMaxSize; size++)
+    {
+        Message *         message = instance->Get<Ip6::Ip6>().NewMessage(sizeof(Ip6::Icmp::Header));
+        Ip6::Icmp::Header icmp6Header;
+        Ip6::MessageInfo  messageInfo;
+
+        VerifyOrQuit(message != nullptr, "Ip6::NewMesssage() failed");
+        SuccessOrQuit(message->SetLength(size), "Message::SetLength() failed");
+
+        // Write ICMP6 header with a random payload.
+
+        Random::NonCrypto::FillBuffer(reinterpret_cast<uint8_t *>(&icmp6Header), sizeof(icmp6Header));
+        icmp6Header.SetChecksum(0);
+        message->Write(0, sizeof(icmp6Header), &icmp6Header);
+
+        if (size > sizeof(icmp6Header))
+        {
+            uint8_t  buffer[kMaxSize];
+            uint16_t payloadSize = size - sizeof(icmp6Header);
+
+            Random::NonCrypto::FillBuffer(buffer, payloadSize);
+            message->Write(sizeof(icmp6Header), payloadSize, &buffer[0]);
+        }
+
+        SuccessOrQuit(messageInfo.GetSockAddr().FromString(kSourceAddress), "FromString() failed");
+        SuccessOrQuit(messageInfo.GetPeerAddr().FromString(kDestAddress), "FromString() failed");
+
+        // Verify that the `Checksum::UpdateMessageChecksum` correctly
+        // updates the checksum field in the ICMP6 header on the message.
+
+        Checksum::UpdateMessageChecksum(*message, messageInfo.GetSockAddr(), messageInfo.GetPeerAddr(),
+                                        Ip6::kProtoIcmp6);
+
+        VerifyOrQuit(message->Read(message->GetOffset(), sizeof(icmp6Header), &icmp6Header) == sizeof(icmp6Header),
+                     "Message::Read() failed");
+        VerifyOrQuit(icmp6Header.GetChecksum() != 0, "Failed to update checksum");
+
+        // Verify that the calculated ICMP6 checksum is valid.
+
+        VerifyOrQuit(CalculateChecksum(messageInfo.GetSockAddr(), messageInfo.GetPeerAddr(), Ip6::kProtoIcmp6,
```

Comment:
Can we add some negative test cases to validate that the checksum catches corrupt data?

Reply:
Good idea. Will add it in next push.

Reply:
Added in new pushed commit.